MaxMind.MinFraud/Request/TransactionReport.cs

```diff
@@ -58,23 +61,27 @@ public sealed class TransactionReport
         /// <param name="maxmindId">A unique eight character string identifying
         ///     a minFraud Standard or Premium request. These IDs are returned
         ///     in the <c>maxmindID</c> field of a response for a successful
-        ///     minFraud request. This field is not required, but you are
-        ///     encouraged to provide it, if possible.</param>
+        ///     minFraud request. This field is not required if you provide at
+        ///     least one of the transaction's <c>ipAddress</c>,
+        ///     <c>minfraudId</c>, or <c>transactionId></c>. You are encouraged
+        ///     to provide it, if possible.</param>
         /// <param name="minfraudId">A UUID that identifies a minFraud Score,
         ///     minFraud Insights, or minFraud Factors request. This ID is
         ///     returned at <c>/id</c> in the response. This field is not
-        ///     required, but you are encouraged to provide it if the request
-        ///     was made to one of these services.</param>
+        ///     required if you provide at least one of the transaction's
+        ///     <c>ipAddress</c>, <c>maxmindId</c>, or <c>transactionId></c>.
+        ///     You are encouraged to provide it, if possible.</param>
         /// <param name="notes">Your notes on the fraud tag associated with the
         ///     transaction. We manually review many reported transactions to
         ///     improve our scoring for you so any additional details to help
         ///     us understand context are helpful.</param>
         /// <param name="transactionId">The transaction ID you originally passed
-        ///     to minFraud. This field is not required, but you are encouraged
-        ///     to provide it or the transaction's <c>>maxmindId</c> or
-        ///     <c>minfraudId</c>.</param>
+        ///     to minFraud. This field is not required if you provide at least
+        ///     one of the transaction's <c>ipAddress</c>, <c>maxmindId</c>,
+        ///     or <c>minfraudId></c>. You are encouraged to provide it, if
+        ///     possible.</param>
         public TransactionReport(
-            IPAddress ipAddress,
+            IPAddress? ipAddress,
```

Comment:
It is not clear to me how to disambiguate the constructors that way. If I keep the original constructor and add an identical constructor except `ipAddress` is optional, the compiler will give a "The call is ambiguous between the following methods or properties" error. Even if it could disambiguate it seems like it would give the deprecation error anytime the ipAddress was passed, similar to the following:
```
# `!` marks a named parameter as required, which are otherwise optional without it
$ raku -e 'multi sub foo(:$ipaddress!, :$tag!) { say "deprecated constructor" }; multi sub foo(:$ipaddress, :$tag) { say "new constructor" }; foo(tag => 42); foo(ipaddress => 42, tag => 42)'
new constructor
deprecated constructor
```
Is there some pattern for working around this? It sort of seems like what we want is to deprecate passing ipAddress as the first positional parameter.

Reply:
Ah, yeah, that is a bit annoying. I was thinking something like:

```csharp
        public TransactionReport(
            TransactionReportTag tag,
            string? chargebackCode = null,
            IPAddress? ipAddress = null,
            string? maxmindId = null,
            Guid? minfraudId = null,
            string? notes = null,
            string? transactionId = null
        )
```

The constructors aren't necessarily ambiguous, but when the callers specify all of the parameters by name, they are. I think you could work around this by changing the existing constructor to something like:

```csharp
        [Obsolete]
        public TransactionReport(
            IPAddress? ipAddress,
            TransactionReportTag tagObsolete,
            string? chargebackCode = null,
            string? maxmindId = null,
            Guid? minfraudId = null,
            string? notes = null,
            string? transactionId = null
        )
```

It is a bit ugly, but it should solve the issue. Given that we are doing a major version release (or so it seems), we could forgo this, but I think it would be nice to give people a few releases to update their code before dropping the old construcotr.

Reply:
If I understand correctly the thought was the ordering of the parameters (which otherwise appear identical minus what I assume is copy-pasta `tagObsolete`) between the two candidate methods would disambiguate them. I don't think that'll work though:

> Type 'TransactionReport' already defines a member called 'TransactionReport' with the same parameter types

It appears C# does not take ordering into account for method signatures 🤔

Reply:
Yes, it does take ordering into account if you pass ordered parameters. If you pass parameters by name, it does not take the order into account. That latter is the issue here. `tagObsolete` was intentional as it changes the named parameters between the two constructors and tag will always be passed. As such, we continue support anyone who was using positional parameters while anyone using the named parameters would use the new constructor.

Reply:
Ah sorry, I mistakenly had a third candidate of that function in my code that was causing that error 🤦‍♂️